### PR TITLE
FunctionNodes as implementation details

### PR DESCRIPTION
**Pull request by hvy**
This PR hides all `FunctionNode` implementations from `functions/__init__.py` since these are considered implementation details and should not "be visible" to the user.

Some exceptions are `FunctionNode`s with states that are exposed through APIs such as the indices for the max pooling classes and the random states in dropout and gaussian noise.

Tests are modified accordingly to not rely on these classes directly. There are few exceptions including the the ones above.

- [x] Remove FunctionNode imports from `functions/__init__`
- [x] Do not use FunctionNodes classes directly in tests if possible
- ~[ ] Remove `FunctionNode`'s `__init__`'s default arguments~ (Maybe better done in a separate PR)

_Note: Reviewing commit by commit is probably much easier than looking at the total diff._

### CODE REVIEW

**Comment by hvy** [comment]
Rebased on the current master. Need to fix newly introduced FunctionNodes.

- functions.activations.Softplus
- functions.arrays.Permutate
- functions.math.exponential_m1.Expm1

**Comment by hvy** [comment]
Fixed the above mentioned 3 FunctionNodes.

**Comment by hvy** [comment]
Ready for review.

**Comment by hvy** [comment]
I'll fix the broken commit history.

**Comment by hvy** [comment]
@kmaehashi Rebased and fixed the commit history. Could you please take at a look?

File: tests/chainer_tests/functions_tests/pooling_tests/test_spatial_pyramid_pooling_2d.py [hunk 1]
```diff
@@ -190,11 +190,6 @@ def check_invalid_poolings(self):
         with self.assertRaises(ValueError):
             functions.spatial_pyramid_pooling_2d(self.v, 3, pooling='avg')
 
-        with testing.assert_warns(DeprecationWarning), \
```
Comment (kmaehashi) [inline tests/chainer_tests/functions_tests/pooling_tests/test_spatial_pyramid_pooling_2d.py]: I think it makes sense to leave this test.

File: chainer/functions/__init__.py [hunk 2]
```diff
@@ -179,138 +113,87 @@
...
-from chainer.functions.math.trigonometric import Arctan2  # NOQA
 from chainer.functions.math.trigonometric import cos  # NOQA
-from chainer.functions.math.trigonometric import Cos  # NOQA
 from chainer.functions.math.trigonometric import sin  # NOQA
-from chainer.functions.math.trigonometric import Sin  # NOQA
 from chainer.functions.math.trigonometric import tan  # NOQA
-from chainer.functions.math.trigonometric import Tan  # NOQA
 
 from chainer.functions.noise.dropout import dropout  # NOQA
 from chainer.functions.noise.dropout import Dropout  # NOQA
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/functions/__init__.py]: Those are left on purpose, as they're stateful. For `Dropout` and `Gaussian` we sometimes want to reuse their random state/noise. As for the pooling functions, we similarly want to extract the resulting indices.

File: chainer/functions/pooling/upsampling_2d.py [hunk 3]
```diff
@@ -185,8 +185,10 @@ def upsampling_2d(
         ``chainer.config.use_cudnn`` flag when you perform
         :meth:`~chainer.functions.max_pooling_2d` function which will make a
         pooling indicies for this :meth:`~chainer.functions.upsampling_2d`.
-        It is because :attr:`~chainer.functions.MaxPooling2D.indexes` is never
-        created and stored in the :attr:`~chainer.functions.MaxPooling2D`
+        It is because
+        :attr:`~chainer.functions.pooling.max_pooling_2d.MaxPooling2D.indexes`
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/functions/pooling/upsampling_2d.py]: Yes unfortunately... PyTorch's `max_pool**`, takes a boolean flag by the way and returns either the output or a pair of (output, indices) depending on the flag. That'd be another way to do it.

File: chainer/functions/pooling/average_pooling_2d.py [hunk 4]
```diff
@@ -188,7 +188,8 @@ def backward(self, indexes, grad_outputs):
 def average_pooling_2d(x, ksize, stride=None, pad=0):
     """Spatial average pooling function.
 
-    This function acts similarly to :class:`~functions.Convolution2D`, but
+    This function acts similarly to
+    :class:`~functions.connection.convolution_2d.Convolution2D`, but
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/functions/pooling/average_pooling_2d.py]: Thanks, fixed.

File: tests/chainer_tests/functions_tests/pooling_tests/test_spatial_pyramid_pooling_2d.py (hunk 1 above)
Comment (hvy) [reply tests/chainer_tests/functions_tests/pooling_tests/test_spatial_pyramid_pooling_2d.py]: Added the test back.

**Comment by hvy** [comment]
@kmaehashi Is there an already existing upgrade guide for the next release?

**Comment by hvy** [comment]
Documented in upgrade guide.

**Comment by kmaehashi** [comment]
I discussed offline with @hvy -san and would like to propose following spec changes within this PR.

Proposal 1
==========

Remove all FunctionNode classes except `MaxPooling2D` from `chainer.functions.*` namespace.
(`MaxPooling2D` is left because it is a documented API. See Proposal 2.)

However, some FunctionNode classes (`Dropout`, `Gaussian`, `MaxPoolingND`) have states (e.g., mask in dropout).
The concern regarding this proposal is that some users may directly accessing such states from instance. (e.g. `chainer.functions.Dropout().mask`).
This proposal will break compatibility for such users, but I assume majority of users are not using classes in such way.

Proposal 2
==========

Currently users of `F.upsampling_2d` are required to directly work with `MaxPooling2D` class (instead of wrapper function, `F.max_pooling_2d`).
(See example in https://docs.chainer.org/en/latest/reference/generated/chainer.functions.upsampling_2d.html?highlight=MaxPooling2D for details)

```
p = F.MaxPooling2D(2, 2)
with chainer.using_config('use_cudnn', 'never'):
    pooled_x = p.apply((x,))[0]
upsampled_x = F.upsampling_2d(pooled_x, p.indexes, p.kh, p.sy, p.ph, x.shape[2:])
```

This is not good as it suggests users to use private API. I'd like to change the API as follows:

```
pooled_x, indices = F.max_pooling_2d(x, 2, 2, return_indices=True)
upsampled_x = F.upsampling_2d(pooled_x, indices, 2, 2, 0, x.shape[2:])
```

In other words, change `F.max_pooling_2d` like as follows:

- Introduce `return_indices=False` keyword-only argument.
- If `return_indices` is True, disable cuDNN and return index (p.indexes).
- If `return_indices` is False, behave the same way as now.
- Rename `indexes` to `indices` (to be natural English).

**Comment by hvy** [comment]
To Add, Proposal 2 is similar to PyTorch as mentioned here https://github.com/chainer/chainer/pull/4421#discussion_r193667291 and looks like a clean API to me.

**Comment by toslunar** [comment]
Could you make a new PR for Proposal 2?  It seems beneficial regardless of hiding `FunctionNode`s

**Comment by kmaehashi** [comment]
As for proposal 2, I can work on it once it has approved by team member.

**Comment by niboshi** [comment]
We can deprecate `FunctionNode` classes instead of removing them, by inheriting (new) private `FunctionNode` classes. Did you consider this scheme?

E.g.

```
class _Some(FunctionNode):
    ...

class Some(_Some):
    def __init__(self, *args, **kwargs):
        <emit warning>
        super(*args, **kwargs)

def some(...):
    return _Some().apply((...))[0]
```

**Comment by hvy** [comment]
I agree that that'd be more user friendly (I assume you are talking about all classes and not only the ones discussed above). In that case, what was your thought on when to remove them?

**Comment by kmaehashi** [comment]
My assumption is it is safe to merge this as most users already rewrote their direct use of `F.Class` to `F.wrapper` after Function/FunctionNode migration.
https://docs.chainer.org/en/latest/upgrade.html#introduction-of-new-style-functions

**Comment by niboshi** [comment]
@hvy Yes, this is about all the functions. We can remove the aliases after 1 or 2 major releases.
@kmaehashi 
I see, that's a good point. 
But it's been 10 month since v3 beta. That may (or may) not be short enough to be neglected.

**Comment by niboshi** [comment]
Maybe an error rather than a warning is better.

**Comment by kmaehashi** [comment]
After the offline discussion:
* As for proposal 1, we agreed to just remove exposed classes from `chainer.functions`, as proposed in this PR. (It is difficult to display friendly messages for all users, e.g., users who traverses the computational graph and doing sth like `isinstance(xxx, chainer.functions.ReLU)`.)
* As for proposal 2, we'll improve APIs to allow users to work with states without accessing FunctionNode classes. (e.g., #4890)

**Comment by kmaehashi** [comment]
Blocked by #4890. (more to come)

**Comment by hvy** [comment]
Removed Dropout, Gaussian, MaxPooling{2D, N} too. Will rebase/merge this PR once Proposal 2 is finished for all of these functions.

**Comment by kmaehashi** [comment]
I've finished implementing Proposal 2.

> Removed Dropout, Gaussian, MaxPooling{2D, N} too.

`MaxPooling2D` was a documented API, so I think it's safe to keep it until v6. How about moving it to the last line of `chainer/functions/__init__.py` to separate from other `F.func`s, and add some comments (e.g., will be removed in v6)?

**Comment by hvy** [comment]
Thanks a lot. I see, but can we in that case raise an appropriate deprecation warning?

**Comment by kmaehashi** [comment]
I think there's no way to safely do that (as we discussed offline). Let's document it here.
https://github.com/chainer/chainer/pull/4421/files#r193370068

**Comment by hvy** [comment]
Hm that's what I thought. I don't think it makes a significant difference whether we remove it in this major release or in the next in that case. Please correct me if I am wrong.

**Comment by kmaehashi** [comment]
After v5 release, users of v4 may feel "F.MaxPooling2D suddenly disappeared although I was using documented API."

How about updating documents here in separate PR and then backport it to v4?
https://github.com/chainer/chainer/pull/4421/files#r193370068

**Comment by hvy** [comment]
Thanks for the suggestion, but this PR is not going to be backported?

**Comment by kmaehashi** [comment]
I mean:
* Do not fix documentation of `chainer/functions/pooling/upsampling_2d.py` in this PR. Do not backport this PR.
* Create new PR to fix the documentation (to suggest use of `F.max_pooling_2d(return_indices=True)`. Then backport the PR.

**Comment by hvy** [comment]
Sorry I now see what you mean. I think that's an acceptable solution. I'll create a PR to fix the documentation once https://github.com/chainer/chainer/pull/4890 is merged.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by kmaehashi** [comment]
@hvy Could you resolve conflict? (Sorry I had to merge this quickly...)

**Comment by hvy** [comment]
@kmaehashi Done!

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by kmaehashi** [comment]
Thanks! Travis failure seems not related to this PR.
I'll merge this after Jenkins pass.

**Comment by kmaehashi** [comment]
LGTM! Test failures are not related to this PR.
